### PR TITLE
Support for Playlist Tags EXT-X-PLAYLIST-TYPE & EXT-X-STREAM-INF

### PR DESCRIPTION
**Pull request by JungleGenius**
Hi, I needed to parse and dump playlists so I added EXT-X-PLAYLIST-TYPE & EXT-X-STREAM-INF support to your project.  I hope I extended it in a fashion that fits with the overall pattern you were using. I did not add all the fields in the RFC for EXT-X-STREAM-INF, but I added the required and most popular ones.

### CODE REVIEW

**Comment by JungleGenius** [comment]
FYI that StyleCI analysis tool just deleted the escape characters in the RegEx needed to parse the codecs field in my pull request.  Stupid tool.

**Comment by chrisyue** [comment]
Hello @JungleGenius . thanks for this PR 👍 

However, I still have no good idea to handle tags in master playlist in PHPM3u8 2.0, like `#EXT-X-STREAM-INF`. I think it's wired to combine master playlist and media playlist into one. (also `#EXT-X-STREAM-INF` tag must have a `URI` as RFC mentioned, which I have no idea which class it should belong to).

I suggest in this PR, support the `EXT-X-PLAYLIST-TYPE` only.

BTW, I've already start 3.0, which will solve the "master playlist" problem.

**Comment by jamesgraham** [comment]
It would be great to get the `EXT-X-PLAYLIST-TYPE` supported 👍 

@chrisyue do you know when we might get v3 merged?

**Comment by JungleGenius** [comment]
Well I guess its officially forked then. ;-)

BTW, Its not "wired to combine master playlist and media playlist into one" as the the Playlist and Media Playlist are mutually exclusive per the RFC.

"A Playlist is a Media Playlist if all URI lines in the Playlist
   identify Media Segments.  A Playlist is a Master Playlist if all URI
   lines in the Playlist identify Media Playlists.  A Playlist MUST be
   either a Media Playlist or a Master Playlist; all other Playlists are
   invalid."

"Each line is a URI, blank, or starts with the character '#'.  Blank lines are ignored."

So the existing URI model attribute is now being used for holding the URI value for both Media Segments and Media Playlists depending on the M3U8 file type that gets parsed. This is inline with the RFC so I don't see an issue. 

Right now it parses both Playlist and Media Playlist files generated by AWS Elastic Transcoder and dumps them without an issue so it works for our use case.

**Comment by chrisyue** [comment]
@jamesgraham well.. it's hard to say when the 3.0 branch could be merged, I am getting busy with other things recently, but I'll try to finish it in a month.

It'll be much cooler than 2.0 I think. you could check the `feature/3.0` branch to see the changes ;)

**Comment by chrisyue** [comment]
@JungleGenius Yes It could be done in 2.x indeed. 

I think it's much clear to separate the `M3u8` playlist model into two `MediaPlaylist` and `MasterPlaylist` (which is what I am trying to do in `feature/3.0`  #27), also the same case about "stream infs" and "media segments". (I think the "Segment(s)" should better have a more proper name after adding "stream inf" tag support).

However, before releasing 3.0, it seems like this this the only way to have "stream inf tag" supported. I'll try to fix some issues and merge this PR ASAP.

File: src/Segment.php [hunk 1]
```diff
@@ -27,6 +28,7 @@ public function __construct($m3u8Version = null)
         $this->extinfTag = new Tag\ExtinfTag($m3u8Version);
         $this->byteRangeTag = new Tag\ByteRangeTag();
         $this->discontinuityTag = new Tag\DiscontinuityTag();
+        $this->streamTags = new StreamTags();
```
Comment (chrisyue) [inline src/Segment.php:31]: according to 'https://tools.ietf.org/html/rfc8216#section-4.3.4.2', one stream inf tag should have one URI part. so I think `StreamTags` does not belong to a `Segment`, `StreamTag` itself is a `Segment`.

```php
$this->streamTag = new StreamTag();
```

Yes, `Segment` does not sound proper, but we can just define 'Segment' as 'a URI with several tags with it', no matter it's Media or Master playlist.

File: src/StreamTags.php [hunk 2]
```diff
@@ -0,0 +1,109 @@
+<?php
```
Comment (chrisyue) [inline src/StreamTags.php:1]: If you agree that a `StreamTag` is a `Segment`, this class is no longer necessary.

File: src/Tag/DiscontinuitySequenceTag.php [hunk 3]
```diff
@@ -33,11 +33,14 @@ public function getDiscontinuitySequence()
 
     public function dump()
     {
+        if (empty($this->discontinuitySequence) && $this->discontinuitySequence !== '0') {
```
Comment (chrisyue) [inline src/Tag/DiscontinuitySequenceTag.php:36]: I think `null === $this->discontinuitySequence` is enough. if someone doesn't need `discontinuitySequence`, he/she should not call the setter.

File: src/Tag/MediaSequenceTag.php [hunk 4]
```diff
@@ -33,11 +33,14 @@ public function getMediaSequence()
 
     public function dump()
     {
+        if (empty($this->mediaSequence) && $this->mediaSequence !== '0') {
```
Comment (chrisyue) [inline src/Tag/MediaSequenceTag.php:36]: same as `discontinuitySequence`

File: src/Tag/PlaylistTypeTag.php [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+ * file that was distributed with this source code.
+ */
+
+namespace Chrisyue\PhpM3u8\Tag;
+
+class PlaylistTypeTag extends AbstractTag
+{
+    use SingleValueTagTrait;
+
+    private $playlist_type;
```
Comment (chrisyue) [inline src/Tag/PlaylistTypeTag.php:18]: PHP M3u8 adopts "camelCase property" rule

```
private $playlistType
```

File: src/Tag/StreamTag.php [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Chrisyue\PhpM3u8\Tag;
+
+class StreamTag extends AbstractTag
+{
+    use SingleValueTagTrait;
```
Comment (chrisyue) [inline src/Tag/StreamTag.php:16]: why not use `AttributesValueTagTrait`

https://github.com/chrisyue/php-m3u8/blob/develop/src/Tag/AttributesValueTagTrait.php

you could check `KeyTag` class as an example.

File: src/Tag/TargetDurationTag.php [hunk 7]
```diff
@@ -33,11 +33,14 @@ public function getTargetDuration()
 
     public function dump()
     {
+        if (empty($this->targetDuration) && $this->targetDuration !== '0') {
```
Comment (chrisyue) [inline src/Tag/TargetDurationTag.php:36]: same as `discontinuitySequence`

**Comment by jamesgraham** [comment]
Thanks @chrisyue I'm happy to help test when the time comes, much of the new features we implemented manually at the moment. Great library 👍

**Comment by chrisyue** [comment]
@JungleGenius 
I've create a new pull request #33.
If you agree with the changes, I'll update readme and merge it ASAP.

**Pull request closed without merging**